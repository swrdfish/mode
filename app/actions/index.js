export const addMessage = (uid, text, isMine) => ({
  type: 'ADD_MSG',
  uid,
  text,
  isMine
})

export const deleteMessage = (id) => ({
  type: 'DEL_MSG',
  id
})

export const setContactFilter = (username) => ({
  type: 'SET_CONTACT_FILTER',
  username
})

export const addUser = (userData) => ({
  type: 'ADD_USER',
  userData
})

export const removeUser = (userData) => ({
  type: 'DEL_USER',
  userData
})

export const changeUser = (userData) => ({
  type: 'CHG_USER',
  userData
})

export const login = (ref, username, uid, publicIp) => ({
  type: 'LOGIN',
  ref,
  username, 
  uid,
  publicIp
})

export const chat = (uid) => ({
  type: 'CHAT_USER',
  uid
})

export const notify = (message, type) => ({
  type: 'ADD_NOTIFICATION',
  notificationType: type,
  message
})

export const unnotify = (index) => ({
  type: 'DEL_NOTIFICATION',
  index
})

export const joinRoom = (connectionManager, roomRef) => ({
  type: 'JOIN_ROOM',
  connectionManager,
  roomRef
})