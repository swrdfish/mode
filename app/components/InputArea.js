import React from 'react'


class InputArea extends React.Component {
	render() {
		return (
			<div className="input-area-wrapper">
				<input type="text" placeholder="Type message" className="input-area-inputfiled" />
				<i className="fa fa-paperclip" id="attach-button" aria-hidden="true"></i>
			</div>
		)
	}
}

export default InputArea