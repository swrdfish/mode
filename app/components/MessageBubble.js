import React from 'react'
import AutoLinker from 'autolinker'
import escapeHTML from 'escape-html'


class MessageBubble extends React.Component {
    constructor(props) {
        super(props)
    }

    render() {
        let item = this.props.item
        let className = item.isMine?'message-bubble mine':'message-bubble'
        let downloadButtonClass = item.isMine?"":"fa fa-download"
        let pretext = item.isMine?"Sent ":"Download "
        let text

        let itemRendered
        if(item.type == "file") {
            if(item.body.count > 1) {
                text = (
                    <a href="#" onClick={this.props.onDownload} data-id={item.body.id}>
                        {item.body.count} files&nbsp;&nbsp;&nbsp;
                        <span className={downloadButtonClass + " download-button"}/>
                    </a>
                )
            } else {
                text = (
                    <a href="#" onClick={this.props.onDownload} data-id={item.body.id}>
                        {item.body.filenames[0]}&nbsp;&nbsp;&nbsp;
                        <span className={downloadButtonClass + " download-button"}/>
                    </a>
                )
            }

            if(item.isMine) {
                text = "" + item.body.count
                text += item.body.count > 1?" files":" file"
            }
            className += " file"
            itemRendered = (
                <div className='message-bubble-wrapper'>
                    <div className={className}>
                        <span>
                            {pretext}&nbsp;&nbsp;{text}               
                        </span>
                        <span className='time-stamp'>{item.timeStamp.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true })}</span>
                    </div>
                </div>
            )
        } else {
            let messageBody = AutoLinker.link(escapeHTML(item.body), {className: "message-bubble-link"})
            itemRendered = (
                <div className='message-bubble-wrapper'>
                    <div className={className}>
                        <span className='message-bubble-body' dangerouslySetInnerHTML={{ __html: messageBody }}></span>
                        <span className='time-stamp'>{item.timeStamp.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true })}</span>
                    </div>
                </div>
            )
        }
        return itemRendered
    }
}

export default MessageBubble