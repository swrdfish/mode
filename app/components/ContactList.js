import React from 'react'
import { connect } from 'react-redux'
import fuzzysearch from 'fuzzysearch'

import ContactProfilePicture from './ContactProfilePicture.js'
import { chat } from '../actions'


class ContactList extends React.Component {
    constructor(props){
        super(props)
        this.state = {
            selected: ""
        }
        this.handleClick = this.handleClick.bind(this)
    }

    handleClick(e) {
        let selected = e.currentTarget.getAttribute("data-id")
        if (selected) {
            this.setState({selected})
            this.props.dispatch(chat(selected))
            this.props.room.connectionManager.getConnection(selected)
        }
    }

    render() {  
        let userList = this.props.userList || []
        let userToDisplay = userList.map(user => ({
                "name": user.name? user.name : user.localIP,
                "id": user.uid,
                "unseen": user.unseen
            })
        )

        if(this.props.searchText !== ''){
            userToDisplay = userToDisplay.filter(contact => (fuzzysearch(this.props.searchText, contact.name.toLowerCase())))
        }

        let contacts = userToDisplay.map(contact => {
            let className = this.state.selected === contact.id ? 'contact-list-item selected' : 'contact-list-item'
            let unseen = contact.unseen?(
                <div className="unseen-count-wrapper">
                    <span className="unseen-count">{contact.unseen}</span>
                </div>
            ) : ''
            return (
                <div className={className} key={contact.id} data-id={contact.id} onClick={this.handleClick} >
                    <ContactProfilePicture profilePicture="/static/img/default_contact.jpg" />
                    <div className="contact-name">
                        {contact.name}
                    </div>
                    {unseen}
                </div>
            );
        }, this);

        return (
            <div>
                {(contacts.length > 0)? contacts : (<div className="contact-not-found">No contacts found.</div>)}
            </div>
        )
    }
}


const mapStateToProps = ({userList, contactFilter, room, dispatch}) => ({
    searchText: contactFilter,
    userList,
    room,
    dispatch
})

let ConnectedContactList = connect(mapStateToProps)(ContactList) 
export default ConnectedContactList