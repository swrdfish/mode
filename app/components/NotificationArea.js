import React from 'react'
import { connect } from 'react-redux'
import {unnotify} from '../actions'


class NotificationArea extends React.Component {
  constructor(props){
    super(props)
    this.handleClearNotification = this.handleClearNotification.bind(this)
  }

  handleClearNotification(event) {
    this.props.dispatch(unnotify(event.target.getAttribute('data-notification-idx')))
  }

  render() {
    let displayNotifications = this.props.notifications.map((notify, index) => {
      return (
        <div className={notify.type == 'error'? 'notification error': 'notification'} key={index}>
          {notify.msg}
          <i className="fa fa-close clear-notification" aria-hidden="true" data-notification-idx={index} onClick={this.handleClearNotification}></i>
        </div>
      )
    })

    return (
      <div className="notification-area">
        {displayNotifications}
      </div>
    )
  }
}

const mapStateToProps = ({notifications, dispatch}) => ({
  notifications,
  dispatch
})

let ConnectedNotificationArea = connect(mapStateToProps)(NotificationArea) 
export default ConnectedNotificationArea;