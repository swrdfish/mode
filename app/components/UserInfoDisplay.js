import React from 'react';
import UserProfilePicture from './UserProfilePicture'
import UserName from './UserName'

import { connect } from 'react-redux'
import { setContactFilter, setUsername } from '../actions'


class UserInfoDisplay extends React.Component {
  
    constructor(props) {
        super(props)
        this.onUpdateUsername = this.onUpdateUsername.bind(this)
    }

    onUpdateUsername(username) {
        let usersRef = this.props.room.roomRef.child('users')
        let userRef = usersRef.child(this.props.auth.uid)
        userRef.child('username').set(username)
        this.props.dispatch(setUsername(username))
    }

	render() {
        let username = this.props.auth.username || this.props.auth.localIP
        let presence = (this.props.auth.uid !== undefined)? "presence-active" : "presence-inactive"
		return (
			<div className="user-info-wrapper">
				<UserProfilePicture />
				<UserName value={username} onUpdateUsername={this.onUpdateUsername}/>
                <span className={presence} />
			</div>
		)
	}
}


const mapStateToProps = ({auth, room, dispatch}) => ({
  auth,
  room,
  dispatch
})

let ConnectedUserInfoDisplay = connect(mapStateToProps)(UserInfoDisplay)
export default ConnectedUserInfoDisplay