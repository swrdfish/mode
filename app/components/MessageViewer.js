import React from 'react'
import MessageBubble from "./MessageBubble"


class MessageViewer extends React.Component {
    render(){
        if(this.props.chatter) {
            let renderedMsg = (<div className='empty-message-view'>Say something!</div>)
            if (this.props.messages) {
                let messages = this.props.messages.reduceRight((arr, item) => {
                    return arr.concat((<MessageBubble item={item} key={item.id} onDownload={this.props.onDownload}/>))
                }, [])
                renderedMsg = (
                    <div className='messages-scrollable'>
                        {messages}
                    </div>
                )
            }

            return (
                <div className='message-view-wrapper'>
                    {renderedMsg}
                </div>
            )
        } else {
            return (
                <div className='message-view-wrapper'>
                    <div className='empty-message-view'>select someone to chat with.</div>
                </div>
            )
        }
    }
}

export default MessageViewer