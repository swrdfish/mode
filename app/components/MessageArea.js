import React from 'react'
import { connect } from 'react-redux'

import InputArea from './InputArea.js'
import MessageViewer from './MessageViewer.js'


class MessageArea extends React.Component {
    render() {
        let connectionManager = this.props.room.connectionManager
        let chatter = this.props.chatArea.uid
        return (
            <div id="message-area" className="col-md-9">
                <MessageViewer connectionManager={connectionManager} chatter={chatter} />
                <InputArea />
            </div>
        )
    }
}

let mapStateToProps = ({chatArea, room}) => ({
  chatArea,
  room
})

const ConnectedMessageArea = connect(mapStateToProps)(MessageArea)
export default ConnectedMessageArea