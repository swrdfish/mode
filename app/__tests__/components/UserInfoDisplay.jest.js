import React from 'react'
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme'
import { UserInfoDisplay } from '../../components/UserInfoDisplay'
import UserProfilePicture from '../../components/UserProfilePicture'
import { setUsername } from '../../actions'

describe('Component: UserInfoDisplay', () => {
    it('should render the component properly', () => {
        const auth = {
            username: 'foobar',
            uid: 1234,
            localIP: '127.0.0.1'
        }
        const wrapper = shallow(<UserInfoDisplay auth={auth} />);
        expect(wrapper.find(".user-info-wrapper")).toHaveLength(1)
        expect(wrapper.find(UserProfilePicture)).toHaveLength(1)
        expect(wrapper.find("span")).toHaveLength(1)
        expect(wrapper.find("UserName")).toHaveLength(1)
    })

    it('should match previous snapshot', () => {
        const auth = {
            username: 'foobar',
            uid: 1234,
            localIP: '127.0.0.1'
        }
        const tree = renderer
            .create(<UserInfoDisplay auth={auth} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('should handle onUpdateUsername properly', () => {
        const auth = {
            username: 'foobar',
            uid: 1234,
            localIP: '127.0.0.1'
        }

        let setFn = jest.fn()
        let userRef = {
            child(id){
                return {
                    set: setFn
                }
            }
        }

        let usersRef = {
            child(id){
                return userRef
            }
        }

        let room = {
            roomRef: {
                child(id){
                    return usersRef
                }
            }
        }

        let dispatch = jest.fn() 
        const wrapper = shallow(<UserInfoDisplay auth={auth} room={room} dispatch={dispatch} />);
        UserInfoDisplay.prototype.onUpdateUsername.call(wrapper.instance(), 'foobar')
        expect(setFn).toHaveBeenCalledWith('foobar')
        expect(dispatch).toHaveBeenCalledWith(setUsername('foobar'))
    })
})