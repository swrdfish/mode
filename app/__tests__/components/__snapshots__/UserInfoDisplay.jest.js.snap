// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: UserInfoDisplay should match previous snapshot 1`] = `
<div
  className="user-info-wrapper"
>
  <div
    className="profile-picture"
  />
  <span
    className="presence-active"
  />
    
  <span
    className="user-name"
    onClick={[Function]}
  >
    foobar
        
    <span
      className="user-name-edit fa fa-pencil"
    />
  </span>
</div>
`;
